infrastructure/src/main/java/org/corfudb/infrastructure/CompactorLeaderServices.java

```diff
@@ -0,0 +1,485 @@
+            LivenessValidatorHelper.clear();
+            String streamName = TableRegistry.getFullyQualifiedTableName(table.getNamespace(), table.getTableName());
+            UUID streamId = CorfuRuntime.getCheckpointStreamIdFromName(streamName);
+            long currentStreamTail = corfuRuntime.getSequencerView()
+                    .getStreamAddressSpace(new StreamAddressRange(streamId, Address.MAX, Address.NON_ADDRESS)).getTail();
+
+            long syncHeartBeat = -1;
+            try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+                syncHeartBeat = txn.getRecord(activeCheckpointsTable, table).getPayload().getSyncHeartbeat();
+                txn.commit();
+            } catch (Exception e) {
+                syslog.warn("Unable to acquire table status");
+            }
+            if (readCache.containsKey(table)) {
+                //previousStatus.first is syncHeartbeat
+                //previousStatus.second is StreamTail
+                Tuple<Long, Long> previousStatus = readCache.get(table).first;
+                if (previousStatus.second < currentStreamTail) {
+                    readCache.put(table, Tuple.of(Tuple.of(previousStatus.first, currentStreamTail), currentTime));
+                } else if (previousStatus.first < syncHeartBeat) {
+                    readCache.put(table, Tuple.of(Tuple.of(syncHeartBeat, previousStatus.second), currentTime));
+                } else if (previousStatus.second == currentStreamTail || previousStatus.first == syncHeartBeat) {
+                    if (currentTime - readCache.get(table).second > timeout && !handleSlowCheckpointers(table)) {
+                        readCache.clear();
+                        return;
+                    }
+                }
+            } else {
+                readCache.put(table, Tuple.of(Tuple.of(syncHeartBeat, currentStreamTail), currentTime));
+            }
+        }
+    }
+
+    private int findCheckpointProgress() {
+        int idleCount = 0;
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            List<TableName> tableNames = new ArrayList<TableName>(txn.keySet(checkpointingStatusTable)
+                    .stream().collect(Collectors.toList()));
+            idleCount = tableNames.size();
+            for (TableName table : tableNames) {
+                CheckpointingStatus tableStatus = txn.getRecord(checkpointingStatusTable, table).getPayload();
+                if (tableStatus.getStatus() != StatusType.IDLE) {
+                    idleCount--;
+                }
+            }
+            syslog.trace("Number of idle tables: {} out of {}", idleCount, tableNames.size());
+            txn.commit();
+        }
+        return idleCount;
+    }
+
+    private void handleNoActiveCheckpointers(long timeout, long currentTime) {
+        CheckpointingStatus managerStatus = null;
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
```

Comment:
If line 284 is true, then this query result is never used. Consider move this query into the else-block.